graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,201 @@
+                final String query = (String) requestMap.getOrDefault("query", "");
+                return execute(executionInput(query, ctx,
+                                              toVariableMap(requestMap.get("variables")),
+                                              (String) requestMap.get("operationName")));
+            }));
+        }
+        if (contentType.is(MediaType.GRAPHQL)) {
+            return HttpResponse.from(request.aggregate().handleAsync((req, thrown) -> {
+                if (thrown != null) {
+                    logger.warn("{} Failed to aggregate a request:", ctx, thrown);
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String query = req.contentUtf8();
+                return execute(executionInput(query, ctx, null, null));
+            }));
+        }
+        return HttpResponse.of(HttpStatus.UNPROCESSABLE_ENTITY,
+                               MediaType.PLAIN_TEXT,
+                               "Could not process GraphQL request");
+    }
+
+    private ExecutionInput executionInput(String query, ServiceRequestContext ctx,
+                                          @Nullable Map<String, Object> variables,
+                                          @Nullable String operationName) {
+        final Builder builder = ExecutionInput.newExecutionInput(query);
+        if (variables != null) {
+            builder.variables(variables);
+        }
+        if (operationName != null) {
+            builder.operationName(operationName);
+        }
+        if (dataLoaderRegistry != null) {
+            builder.dataLoaderRegistry(dataLoaderRegistry);
+        }
+        return builder.context(ctx).build();
+    }
+
+    private HttpResponse execute(ExecutionInput input) {
+        final ExecutionResult executionResult = executionResult(input);
+        return HttpResponse.of(MediaType.JSON_UTF_8, toJsonString(executionResult.toSpecification()));
+    }
+
+    private ExecutionResult executionResult(ExecutionInput input) {
+        try {
+            final CompletableFuture<ExecutionResult> completableFuture = new CompletableFuture<>();
+            graphQL.executeAsync(input).thenAccept(completableFuture::complete);
+            return completableFuture.get();
+        } catch (RuntimeException | ExecutionException | InterruptedException e) {
+            return new ExecutionResultImpl(GraphqlErrorException.newErrorException()
+                                                                .message(e.getMessage())
+                                                                .cause(e)
+                                                                .build());
+        }
+    }
+
+    private static Map<String, Object> toVariableMap(@Nullable Object variables) {
+        if (variables == null) {
+            return ImmutableMap.of();
+        }
+
+        if (variables instanceof Map) {
```

Comment:
Question: `variables` can be an instance of `Map`?

Reply:
Since `variables` are part of the requested data in JSON format, I think that it may not necessarily be a `Map` instance.